Paged file browser

maks opened the pull request:
This signficantly reduces the ram usage when using the import dialog to browse and import samples from the sdcard.

It changes the Import Dialog when built for picotracker to use a new `picoTrackerPagedDir` class that implements a new interface instead of `I_Dir`.  This changes the file system operations to be done in "pages" of files and only stores the indexes of file and subdirectory entries of the currently open directory and only fetches the names of entries on demand in "pages" limited in size to what the dialog UI can display at one time (currently 15).

The `ImportSampleDialog` is also heavily rewritten to make use of the interface and use as little ram as possible.

This doesn't impose any hard limits on the number of files per directory, but it does set initial sizes of max of `96` subdirs and `354` files per subdir so its quite possible that exceeding these limits could cause extra memory allocation to overflow available ram at the moment.

Please note this is "stacked" on top of PR #26 so that PR should be merged before this one is.

Note: these changes do mean that apart from having subdirs listed before files, there is no other sorting of entries in a directory being applied for now.

Also I've kept all my commits in this PR but I would recommend doing a "squash merge" as all these commits don't really need to be included in the main branches history.

Fixes #14

democloid commented:
For starters a couple of comments.
I should put this on some dev standards file or something. Could you use "clang-format" with default options on new files? I've been doing that in all new files and when I modify others, I regret not having done this on the base project before I started the fork.

How useful is the gui debug code enabled with SHOW_MEM_USAGE? Do you think this makes sense rather than debug logs? Not against per se, but maybe if it's something that we want to support for development we should create some generic define GUI_DEBUG? GUI_DEBUG_MESSAGES?

maks commented:
Thanks for taking a look at this so quickly @democloid ! 👍🏻 

Ah thanks for the pointer, I didn't know about clang-format! I lean on dartfmt heavily when writing in Dart so great to know theres a tool for  c++ as well. I'll reformat all the new files with it. Its tempting for me to reformat existing file too as the mix of tabs and spaces drives me nuts, but the downside there is losing `git blame` to be able to quickly see if code was changed for picotracker or is original Lgpt code.

Oops, sorry I left `SHOW_MEM_USAGE` enable by mistake, will fix that 👍🏻 And double oops, forgot to disable uart debug as well.
But in general, yes I found it very useful to have on the screen during debugging as I had *alot* of debug logging enabled as I don't yet have uart for gdb setup so I was doing "print line debugging" and having mem usage on the display vs swapped by lots of other logs did help. So having a more general `#define` would probably be a good idea but for now I can't think of other UI I want to add so maybe just keep it for now until we need something more general?

maks commented:
@democloid I've updated to remove the debug build settings and also reformatted my new files and the 2 main files I changed in the picotracker subdir with clang-format (default settings) but I kept the others as is to try to make it clearer what changes I made as part of this (except for `SelectProjectDialog` where I had already changed some of the whitespace formatting but didnt seem worth it to reformat the whole file with clang-format).

democloid commented on sources/Application/Views/ModalDialogs/SelectProjectDialog.cpp:
```diff
@@ -282,32 +281,27 @@ void SelectProjectDialog::setCurrentFolder(Path &path) {
 		dir->GetContent("*");
-    dir->Sort();
-      
+		// dir->Sort();
```
Should we keep this and define an empty sort method for the moment? (possibly with a log stating sorting is not implemented yet)

I'd leave this file untouched, most changes are unrelated

maks replied:
good point! will do.

maks replied:
actually this is still using the older I_Dir so it can still do the sorting

democloid commented on sources/System/FileSystem/FileSystem.h:
```diff
@@ -81,8 +93,8 @@ class I_File {
 	virtual int Read(void *ptr, int size, int nmemb)=0;
-  virtual int GetC() = 0;
-  virtual int Write(const void *ptr,int size, int nmemb)=0;
+	virtual int GetC() = 0;
```
revert tabs

democloid reviewed (changes requested):
Thanks for this PR! do you have a sense on how much total memory is being used in the worst case scenario?

I still think this should be made static mem, at the end of the day, even if this memory is used only in this part of the code, it's very hard to accurately keep track of how much heap is used by different components in order to effectively share that memory usage. The risks of weird bugs apearing is too high. We can leave that for later tho, as this will make the current situation better anyway and we can convert later.

maks commented on sources/Adapters/picoTracker/filesystem/picoTrackerFileSystem.cpp:
```diff
@@ -3,6 +3,123 @@
+picoTrackerPagedDir::picoTrackerPagedDir(const char *path)
+    : path_{std::string(path)} {
+  fileIndexes_.reserve(354);
+  subdirIndexes_.reserve(96);
```
Yes good point! It is very arbitrary, I just picked numbers that experimentally didnt cause any issues when I tested them. 
But these aren't actually limits per se, this just sets the initial capacity of the vectors, they will dynamically grow if items are added that exceed the current capacity, its just when that happens, that will likely exceed available ram, at least thats available at the moment, without further ram usage optimisations.

maks commented on sources/Adapters/picoTracker/filesystem/picoTrackerFileSystem.cpp:
```diff
@@ -3,6 +3,123 @@
+  char current[MAX_FILENAME_LEN];
+  FsBaseFile file;
+
+  if (startOffset == 0 && (path_ != std::string("/samplelib"))) {
```
oops, yes well spotted, I'll fix this up! 👍🏻

maks commented on sources/Adapters/picoTracker/filesystem/picoTrackerFileSystem.cpp:
```diff
@@ -3,6 +3,123 @@
+#include <string>
+
+// Max filename is actually 256 per FAT std
+static const int MAX_FILENAME_LEN = 128;
```
Yes, I use `MAX_FILENAME_LEN` for reading the filename for example when we need to actually open the file, though I'm no longer passing more than 24 chars out to the callers of this class, ie, SampleImportDialog as it can't actually display longer filenames as you mention.

maks commented on sources/Adapters/picoTracker/filesystem/picoTrackerFileSystem.cpp:
```diff
@@ -3,6 +3,123 @@
+    Trace::Error("Path:%s is not a directory", path_.c_str());
+    return;
+  }
+  static const int MAX_ITEMS = 15;
```
I have a (bad?) habit of trying not to use the preprocessor, but yes it could be a define though I think it would basically come out as the same generated code I think, but I can change it to a define if you prefer that?

maks commented on sources/Adapters/picoTracker/gui/picoTrackerGUIWindowImp.cpp:
Yes sorry that is due to stacking on previous PRs.

maks commented on sources/Application/Model/Project.h:
```diff
@@ -17,7 +17,7 @@
 #define PROJECT_NUMBER "1.0"
 #define PROJECT_RELEASE "r"
-#define BUILD_COUNT "003"
+#define BUILD_COUNT "004"
```
oops yes! I think that got left in by mistake

maks commented on sources/Application/Views/InstrumentView.cpp:
```diff
@@ -274,7 +275,8 @@ void InstrumentView::ProcessButtonMask(unsigned short mask,bool pressed) {
               } else { ;
                 // Go to import sample
-                ImportSampleDialog *isd=new ImportSampleDialog(*this) ;
+                // ImportSampleDialog *isd=new ImportSampleDialog(*this) ;
```
oops yes well spotted! I meant to delete that before commiting.

maks commented on sources/Application/Views/ModalDialogs/ImportSampleDialog.cpp:
yes thanks! will remove.

maks commented on sources/Application/Views/ModalDialogs/PagedImportSampleDialog.cpp:
```diff
@@ -0,0 +1,269 @@
+
+const static char *sampleLib_ = "/samplelib";
+
+static const char *buttonText[3] = {"Listen", "Import", "Exit"};
```
ah sorry, I didn't even notice that both were valid in c++. I'll make it consistent.

maks commented on sources/Application/Views/ModalDialogs/PagedImportSampleDialog.cpp:
I kept it as I still want to have the orig implementation for now to use on desktop platforms as I still have hopes of using this code as the basis for a "modernised" lgpt for Linux desktop & game handhelds.

maks commented on sources/Application/Views/ModalDialogs/PagedImportSampleDialog.cpp:
```diff
@@ -0,0 +1,269 @@
+
+  SamplePool *pool = SamplePool::GetInstance();
+
+#ifdef PICO_BUILD
```
oops I should have removed that, thanks for spotting that 👍🏻

maks commented on sources/Application/Views/ModalDialogs/PagedImportSampleDialog.h:
```diff
@@ -0,0 +1,36 @@
+private:
+  std::vector<FileListItem> fileList_{};
+  int currentSample_;
+  int topIndex_;
```
yes good call! will fix it.

maks commented on sources/Application/Views/ModalDialogs/PagedImportSampleDialog.cpp:
```diff
@@ -0,0 +1,269 @@
+      fullPathStr += currentItem.name;
+      auto fullPath = Path{fullPathStr};
+
+      switch (selected_) {
```
actually I have a follow up PR that gets rid of the buttons and then `selected_` will just go away

maks commented:
@democloid I've hopefully addressed all the feedback, either with code changes or replies, but do let me know if you have followup comments 👍🏻 .
